### PR TITLE
test: Add coverage for dumptxoutset failure robustness

### PR DESCRIPTION
**Pull request by fjahr**
This adds a test that checks that network activity is not suspended if dumptxoutset fails in the middle of its process which is implemented with the `NetworkDisable` RAII class. I would have liked to add coverage for the `TemporaryRollback` RAII class but that seems a lot more tricky since the failure needs to happen at some point after the rollback and on the scale of our test chain here I couldn't find a way to do it yet. This was requested by pablomartin4btc here: https://github.com/bitcoin/bitcoin/pull/30808#pullrequestreview-2280450117. To test the test you can comment out the content of the destructor of `NetworkDisable`.

It also addresses the feedback by ryanofsky to use `std::optional` instead of `std::unique_ptr` for the management of the RAII object: https://github.com/bitcoin/bitcoin/pull/30808#discussion_r1744149228

### CODE REVIEW

File: test/functional/rpc_dumptxoutset.py [hunk 1]
```diff
@@ -60,6 +62,16 @@ def run_test(self):
         assert_raises_rpc_error(
             -8, 'Invalid snapshot type "bogus" specified. Please specify "rollback" or "latest"', node.dumptxoutset, 'utxos.dat', "bogus")
 
+        self.log.info(f"Test that dumptxoutset failure does not leave the network activity suspended")
+        rev_file = node.chain_path / "blocks" / "rev00000.dat"
+        bogus_file = node.chain_path / "blocks" / "bogus.dat"
```
Comment (tdb3) [inline test/functional/rpc_dumptxoutset.py:67]: Could use `node.blocks_path` instead of `node.chain_path` for these two lines.

**Review by tdb3 — COMMENTED** [review]
Approach ACK

Nice addition.  Definitely don't want network broken from a failed RPC.
Keeping the PR scoped to just `NetworkDisable` seems good (more concise to reviewers).

File: test/functional/rpc_dumptxoutset.py (hunk 1 above)
Comment (fjahr) [reply test/functional/rpc_dumptxoutset.py]: done

File: test/functional/rpc_dumptxoutset.py [hunk 2]
```diff
@@ -60,6 +62,16 @@ def run_test(self):
...
+        os.rename(rev_file, bogus_file)
```
[the remark this answers is not in the document]
Comment (fjahr) [reply test/functional/rpc_dumptxoutset.py]: done

**Comment by fjahr** [comment]
> You can also remove UniValue error

Done in the refactoring commit, thanks!

**Review by pablomartin4btc — APPROVED** [review]
cr & tACK c2b779da4e7f1bf1a5c5d67ec94cba3027b42ee7

**Review by tdb3 — APPROVED** [review]
ACK c2b779da4e7f1bf1a5c5d67ec94cba3027b42ee7

**Review by BrandonOdiwuor — APPROVED** [review]
Code Review ACK c2b779da4e7f1bf1a5c5d67ec94cba3027b42ee7

**Review by theStack — APPROVED** [review]
ACK c2b779da4e7f1bf1a5c5d67ec94cba3027b42ee7

Further potential follow-up material (maybe a "good first issue" candidate): could go one step further and even test that if the network was already disabled (via `setnetworkactive False`), it won't be enabled after a `dumptxoutset` call. Right now the following patch passes all tests:
```diff
diff --git a/src/rpc/blockchain.cpp b/src/rpc/blockchain.cpp
index b94b3a64e1..b1267bf808 100644
--- a/src/rpc/blockchain.cpp
+++ b/src/rpc/blockchain.cpp
@@ -2812,9 +2812,7 @@ static RPCHelpMan dumptxoutset()
         // Skip if the network is already disabled because this
         // automatically re-enables the network activity at the end of the
         // process which may not be what the user wants.
-        if (connman.GetNetworkActive()) {
-            disable_network.emplace(connman);
-        }
+        disable_network.emplace(connman);
 
         invalidate_index = WITH_LOCK(::cs_main, return node.chainman->ActiveChain().Next(target_index));
         temporary_rollback.emplace(*node.chainman, *invalidate_index);
```

**Comment by achow101** [comment]
ACK c2b779da4e7f1bf1a5c5d67ec94cba3027b42ee7

**Comment by fjahr** [comment]
> could go one step further and even test that if the network was already disabled (via setnetworkactive False), it won't be enabled after a dumptxoutset call.

Added in #30892

**Pull request merged**